add native quantization support for pixel_shuffle

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74999
* __->__ #68328
* #72575
* #71903
* #69667
* #69601
* #69600
* #69599
* #69598
* #69517

Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

mingfeima committed:
add native quantization support for pixel_shuffle

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3326,6 +3326,10 @@
 
 - func: pixel_unshuffle(Tensor self, int downscale_factor) -> Tensor
 
+- func: quantized_pixel_shuffle(Tensor self, int upscale_factor) -> Tensor
```
Why do you need a new op for this if it is only called from channel_shuffle? This could just be a different dispatch for channel_shuffle below no?

mingfeima replied:
@albanD Hi thanks! 
Updated with dispatch key of `QuantizedCPU` and `CompositeExplicitAutograd`.

If I use `CPU, CUDA` in the dispatch table, XLA test case would fail.

If I use `CompositeImplicitAutograd`, it gave the following error:
```
RuntimeError: 0INTERNAL ASSERT FAILED at "../aten/src/ATen/core/boxing/KernelFunction.cpp":23, please report a bug 
to PyTorch. aten::pixel_shuffle has kernels registered to both CompositeImplicitAutograd and a backend mapped to 
AutogradOther. This makes the backend kernel unreachable; the dispatcher will always prefer the 
CompositeImplicitAutograd lowering (see Note [Ambiguity in AutogradOther kernel]).
```

mingfeima commented:
- [ ] use dispatchkey of QuantizedCPU
- [ ] provide benchmarking data

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima commented:
This is part of effort optimizing one of our customer model **MSRResNet**.

Originally `PixelShuffle` does not have native implementation on `QuantizedCPU` path, which lead to bad performance.  To make sure this kernel have good performance on both single core and multi-core, this patch did parallel on dimension of `{N * H}` (so that N=1 will receive performance benefits using multi-core as well) and used transpose utils from fbgemm.

The transpose is done via two steps as i put in the code comments, the temporary buffer is sized of {c} per thread which should fall in L1 cache for commonly used model parameter configs. Theoretically, it is possible to fuse step1 and step2 into 1 but in that way i can't use the fbgemm transpose util which means too much effort and very little gain. Also even with the current 2-step implementation, the performance is already satisfactory so i decide to stop here.

The following is performance compare with this path on the major module in the model:

* multi core run (Xeon CLX, 20 cores), **10.5x** performance improvement on PixelShuffle:
```bash
(before)
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                      aten::clone        59.41%        1.286s        59.42%        1.286s     321.558ms             4
              aten::pixel_shuffle         0.00%      35.000us        57.20%        1.238s     619.121ms             2
                    aten::reshape         0.00%      78.000us        57.20%        1.238s     309.541ms             4
                quantized::conv2d        27.82%     602.310ms        30.17%     653.064ms     326.532ms             2
                 aten::dequantize        12.19%     263.910ms        12.19%     263.919ms     263.919ms             1
                 aten::contiguous         0.00%      12.000us         2.23%      48.223ms      24.111ms             2
        aten::quantize_per_tensor         0.44%       9.544ms         0.44%       9.544ms       9.544ms             1
    aten::_empty_affine_quantized         0.06%       1.326ms         0.06%       1.326ms     221.000us             6
                      aten::empty         0.06%       1.250ms         0.06%       1.250ms     416.667us             3
                    aten::permute         0.00%      37.000us         0.00%      43.000us      21.500us             2
             aten::_reshape_alias         0.00%      36.000us         0.00%      36.000us      18.000us             2
               aten::_unsafe_view         0.00%      17.000us         0.00%      32.000us      16.000us             2
                       aten::view         0.00%      15.000us         0.00%      15.000us       7.500us             2
                       aten::item         0.00%      11.000us         0.00%      13.000us       6.500us             2
                    aten::qscheme         0.00%      12.000us         0.00%      12.000us       1.000us            12
                    aten::q_scale         0.00%       8.000us         0.00%       8.000us       1.333us             6
                 aten::as_strided         0.00%       6.000us         0.00%       6.000us       3.000us             2
               aten::q_zero_point         0.00%       5.000us         0.00%       5.000us       0.833us             6
        aten::_local_scalar_dense         0.00%       2.000us         0.00%       2.000us       1.000us             2
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 2.165s

(after)
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                quantized::conv2d        59.24%     613.267ms        60.35%     624.691ms     312.346ms             2
                 aten::dequantize        27.00%     279.489ms        27.00%     279.502ms     279.502ms             1
              aten::pixel_shuffle        11.46%     118.615ms        11.46%     118.619ms      59.309ms             2
        aten::quantize_per_tensor         1.19%      12.365ms         1.19%      12.365ms      12.365ms             1
                 aten::contiguous         0.00%      15.000us         0.82%       8.530ms       8.530ms             1
                      aten::clone         0.82%       8.497ms         0.82%       8.515ms       8.515ms             1
                      aten::empty         0.16%       1.662ms         0.16%       1.662ms     554.000us             3
    aten::_empty_affine_quantized         0.12%       1.253ms         0.12%       1.253ms     417.667us             3
                       aten::item         0.00%      12.000us         0.00%      14.000us       7.000us             2
                    aten::q_scale         0.00%       7.000us         0.00%       7.000us       1.400us             5
               aten::q_zero_point         0.00%       5.000us         0.00%       5.000us       1.000us             5
        aten::_local_scalar_dense         0.00%       2.000us         0.00%       2.000us       1.000us             2
                    aten::qscheme         0.00%       2.000us         0.00%       2.000us       0.667us             3
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 1.035s
```

* single core run, **13.3x** performance improvement on PixelShuffle:
```
(before)
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                      aten::clone        71.61%       21.728s        71.61%       21.728s        5.432s             4
              aten::pixel_shuffle         0.00%      32.000us        69.04%       20.950s       10.475s             2
                    aten::reshape         0.00%      57.000us        69.04%       20.950s        5.237s             4
                quantized::conv2d        20.96%        6.361s        23.54%        7.142s        3.571s             2
                 aten::dequantize         7.15%        2.171s         7.15%        2.171s        2.171s             1
                 aten::contiguous         0.00%      14.000us         2.57%     778.569ms     389.284ms             2
        aten::quantize_per_tensor         0.27%      80.611ms         0.27%      80.611ms      80.611ms             1
    aten::_empty_affine_quantized         0.00%       1.338ms         0.00%       1.338ms     223.000us             6
                      aten::empty         0.00%       1.243ms         0.00%       1.243ms     414.333us             3
             aten::_reshape_alias         0.00%      34.000us         0.00%      34.000us      17.000us             2
               aten::_unsafe_view         0.00%      13.000us         0.00%      31.000us      15.500us             2
                    aten::permute         0.00%      25.000us         0.00%      29.000us      14.500us             2
                       aten::view         0.00%      18.000us         0.00%      18.000us       9.000us             2
                       aten::item         0.00%      11.000us         0.00%      12.000us       6.000us             2
                    aten::q_scale         0.00%       8.000us         0.00%       8.000us       1.333us             6
               aten::q_zero_point         0.00%       6.000us         0.00%       6.000us       1.000us             6
                    aten::qscheme         0.00%       4.000us         0.00%       4.000us       0.333us            12
                 aten::as_strided         0.00%       4.000us         0.00%       4.000us       2.000us             2
        aten::_local_scalar_dense         0.00%       1.000us         0.00%       1.000us       0.500us             2
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 30.343s

(after)
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                quantized::conv2d        61.60%        6.450s        62.84%        6.580s        3.290s             2
                 aten::dequantize        20.72%        2.170s        20.72%        2.170s        2.170s             1
              aten::pixel_shuffle        15.63%        1.636s        15.63%        1.636s     818.029ms             2
                 aten::contiguous         0.00%       8.000us         1.22%     127.562ms     127.562ms             1
                      aten::clone         1.22%     127.536ms         1.22%     127.554ms     127.554ms             1
        aten::quantize_per_tensor         0.81%      84.620ms         0.81%      84.620ms      84.620ms             1
    aten::_empty_affine_quantized         0.01%       1.303ms         0.01%       1.303ms     434.333us             3
                      aten::empty         0.01%       1.252ms         0.01%       1.252ms     417.333us             3
                       aten::item         0.00%      11.000us         0.00%      13.000us       6.500us             2
                    aten::q_scale         0.00%       6.000us         0.00%       6.000us       1.200us             5
               aten::q_zero_point         0.00%       5.000us         0.00%       5.000us       1.000us             5
        aten::_local_scalar_dense         0.00%       2.000us         0.00%       2.000us       1.000us             2
                    aten::qscheme         0.00%       1.000us         0.00%       1.000us       0.333us             3
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 10.470s
```

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353157).

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima commented:
The stack is newly rebased!

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -1066,6 +1066,29 @@ def test_channel_shuffle(self, X, groups):
                          msg="torch.nn.functional.channel_shuffle results are off")
 
+    """Tests pixel shuffle operation on quantized tensors."""
+    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=4, max_dims=4,
```
please use simple loops to choose between different options, or just hardcode these parameters, we plan to move away from hypothesis, example: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_op.py#L439

mingfeima replied:
The test case has been fixed, hypothesis is removed.

HDCharles commented on aten/src/ATen/native/quantized/cpu/qpixel_shuffle.cpp:
```diff
@@ -0,0 +1,115 @@
+  int64_t stride_w = channels;
+  int64_t stride_c = S * S;
+  int64_t stride_s1 = S;
+  int64_t stride_s2 = 1;
```
lines 57-59 instantiate unused variables

mingfeima replied:
Updated, unused variables are removed!

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3396,8 +3396,13 @@
 - func: pixel_shuffle(Tensor self, int upscale_factor) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: pixel_shuffle
+    QuantizedCPU: quantized_pixel_shuffle_cpu
```
nit: pixel_shuffle_quantized_cpu

we have aligned on the naming before to ${func_name}_${backend_name}

mingfeima replied:
Updated!

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qpixel_shuffle.cpp:
```diff
@@ -0,0 +1,107 @@
+      self_nhwc.q_zero_point(),
+      MemoryFormat::ChannelsLast);
+
+  auto qx_data = self_nhwc.data_ptr<c10::quint8>();
```
nit: I think we can move reinterpret_cast here, same for qy_data

mingfeima replied:
yes, it should be a static cast

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qpixel_shuffle.cpp:
```diff
@@ -0,0 +1,107 @@
+  // parallel on both n and h dimension, use two steps to do the transformation
+  at::parallel_for(0, nbatch * height, 0, [&](int64_t begin, int64_t end) {
+    // thread local temp buffer
+    std::unique_ptr<c10::quint8 []> buffer(new c10::quint8[channels]);
```
after we move the reinterpret_cast, we can allocate uint8 array directly

mingfeima replied:
updated!

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks great, thanks for addressing the comments!

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353157).

suo commented:
this broke master https://hud2.pytorch.org/pytorch/pytorch/commit/0f8a1fc9ba7927ca7b0af2fbeb2196f147eee0e1

I am reverting

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima commented:
Update: skip test case `test_pixel_shuffle` from test_quantization.py if fbgemm is not supported.

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

mingfeima committed:
Update on "add native quantization support for pixel_shuffle"


Differential Revision: [D33353157](https://our.internmc.facebook.com/intern/diff/D33353157)

[ghstack-poisoned]

jerryzh168 commented:
please make sure ci passes

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3426,8 +3426,13 @@
+    QuantizedCPU: pixel_shuffle_quantized_cpu
 
 - func: pixel_unshuffle(Tensor self, int downscale_factor) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: pixel_unshuffle
```
do we need this?

albanD replied:
If you want to be able to manually define the backward for it, yes.

mingfeima replied:
Ohhh, i see...

jerryzh168 commented on tools/build_variables.bzl:
```diff
@@ -1231,6 +1231,7 @@ aten_native_source_non_codegen_list = [
     "aten/src/ATen/native/quantized/cpu/qmatmul.cpp",
     "aten/src/ATen/native/quantized/cpu/qmul.cpp",
     "aten/src/ATen/native/quantized/cpu/qnormalization.cpp",
+    "aten/src/ATen/native/quantized/cpu/qpixel_shuffle.cpp",
```
another nit: we are trying to aligning the filename under native/quantized/cpu to be the same as the floating point variant, e.g.: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/PixelShuffleKernel.cpp can you change the name as well

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qpixel_shuffle.cpp:
```diff
@@ -0,0 +1,107 @@
+      self_nhwc.q_zero_point(),
+      MemoryFormat::ChannelsLast);
+
+  auto qx_data = self_nhwc.data_ptr<c10::quint8>();
```
does using `uint8_t` instead of `c10::quint8` directly work?

jerryzh168 reviewed (approved):
please make sure the ci is green before landing

jerryzh168 commented:
can you add some descriptions of the change and the performance in the summary?

mingfeima commented:
close this one for now since it is not prioritized.
